extensions/data-workspace/src/dialogs/projectDashboard.ts

```diff
@@ -0,0 +1,124 @@
+
+export class ProjectDashboard {
+
+	private dashboard: azdata.window.ModelViewDashboard | undefined;
+	private modelView: azdata.ModelView | undefined;
+	private projectProvider: IProjectProvider | undefined;
+	private overviewTab: azdata.DashboardTab | undefined;
+
+	constructor(private workspaceService: IWorkspaceService, private treeItem: WorkspaceTreeItem) {
+	}
+
+	public async showDashboard(): Promise<void> {
+		const project = this.treeItem.element.project;
+		this.projectProvider = await this.workspaceService.getProjectProvider(vscode.Uri.file(project.projectFilePath));
+		if (!this.projectProvider) {
+			throw new Error(constants.ProviderNotFoundForProjectTypeError(project.projectFilePath));
+		}
+
+		await this.createDashboard(project.projectFileName);
+		await this.dashboard!.open();
+	}
+
+	private async createDashboard(title: string): Promise<void> {
+		this.dashboard = azdata.window.createModelViewDashboard(title, 'ProjectDashboard', { alwaysShowTabs: false });
+		this.dashboard.registerTabs(async (modelView: azdata.ModelView) => {
+			this.modelView = modelView;
+
+			this.overviewTab = {
+				title: '',
+				id: 'overview-tab',
+				content: this.createContainer(),
+				toolbar: this.createToolbarContainer()
+			};
+			return [
+				this.overviewTab
+			];
+		});
+	}
+
+	private createToolbarContainer(): azdata.ToolbarContainer {
+		const projectActions: (IProjectAction | IProjectActionGroup)[] = this.projectProvider!.projectActions;
+
+		// Add actions as buttons
+		const buttons: azdata.ToolbarComponent[] = [];
+
+		projectActions.forEach(action => {
+			if (this.isProjectAction(action)) {
+				let button = this.createButton(action);
+
+				buttons.push({ component: button });
+			} else {
+				const groupLength = action.actions.length;
+				let currentElement = { count: 0 };
+
+				action.actions.forEach(groupAction => {
+					let button = this.createButton(groupAction);
+
+					buttons.push({ component: button, toolbarSeparatorAfter: this.toolbarSeparator(groupLength, currentElement) });
+				});
+			}
+		});
+
+		return this.modelView!.modelBuilder.toolbarContainer()
+			.withToolbarItems(
+				buttons
+			).component();
+	}
+
+	private isProjectAction(obj: any): obj is IProjectAction {
+		return obj.id !== undefined;
+	}
+
+	private createButton(projectAction: IProjectAction): azdata.ButtonComponent {
+		let button = this.modelView!.modelBuilder.button()
+			.withProperties<azdata.ButtonProperties>({
+				label: projectAction.id,
+				iconPath: projectAction.icon,
+				height: '20px'
+			}).component();
+
+		button.onDidClick(async () => {
+			await projectAction.run(this.treeItem);
+		});
+
+		return button;
+	}
+
+	/**
+	 * Add a toolbar separator at the end of the group actions
+	 * @param length Total number of actions in a group
+	 * @param currentElement Count of current element
+	 */
+	private toolbarSeparator(length: number, currentElement: { count: number }): boolean {
```

Comment:
this is not a reusable method, please remove it and you can do this logic inline, array.foreach not only give you access to the element but also the index: https://www.w3schools.com/jsref/jsref_forEach.asp, you can achieve this using one line of code in createToolbarContainer

Reply:
Thanks for pointing this. This was a much better way of doing the calculation.
I updated the check for adding toolbar separator at the end of the group, but not at the end of the list (if the group is the last one).